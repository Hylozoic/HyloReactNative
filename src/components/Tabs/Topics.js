import React from 'react'
import { Text } from 'react-native'

import WelcomeScene from '../WelcomeScene'
import MenuButton from './MenuButton'
import styles from './styles'

export default class Topics extends React.Component {
  static navigationOptions = ({navigation}) => ({
    headerTitle: <Text style={styles.headerTitle}>Topics</Text>,
    headerLeft: <MenuButton navigation={navigation} />
  })

  render () {
    const { navigation } = this.props
    return <WelcomeScene navigation={navigation} />
  }
}
