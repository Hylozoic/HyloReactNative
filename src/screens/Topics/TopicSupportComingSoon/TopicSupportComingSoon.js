import React from 'react'
import { View, Text, Image } from 'react-native'
import SafeAreaView from 'react-native-safe-area-view'
import LinkButton from 'navigation/linking/LinkButton'
import styles from './TopicSupportComingSoon.styles'

const axolotlImage = require('assets/Axolotyl_Digging.png')

export default function TopicSupportComingSoon ({ style, navigation }) {
  return (
    <SafeAreaView style={[style, styles.container]}>
      <View style={styles.header}>
        <Text style={styles.headerText}>
          We're working on expanding #topics to more places.
        </Text>
      </View>
      <View style={styles.paddedRow}>
        <Text style={styles.bodyText}>
          In the meantime, click a #topic from an individual community to see posts from that community.
        </Text>
      </View>
      <Image style={styles.image} resizeMode='contain' source={axolotlImage} />
      <View style={styles.paddedRow}>
        <LinkButton to='/' style={styles.goBackButton}>
          <Text style={styles.goBackButtonText}>Go Back</Text>
        </LinkButton>
      </View>
    </SafeAreaView>
  )
}
