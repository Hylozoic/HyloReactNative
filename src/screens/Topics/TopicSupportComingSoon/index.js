import component from './TopicSupportComingSoon'

export default component
