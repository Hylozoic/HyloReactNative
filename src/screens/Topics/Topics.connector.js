import { connect } from 'react-redux'
import getCurrentGroup from 'store/selectors/getCurrentGroup'
import fetchGroupTopics, { FETCH_GROUP_TOPICS } from 'store/actions/fetchGroupTopics'
import {
  getGroupTopics, presentGroupTopic, setTopicSubscribe, getTerm, setTerm
} from './Topics.store'
import { get } from 'lodash/fp'

export function mapStateToProps (state, props) {
  const searchTerm = getTerm(state)
  const group = getCurrentGroup(state, props)
  const pending = state.pending[FETCH_GROUP_TOPICS]
  const queryResultParams = {
    id: get('id', group),
    autocomplete: ''
  }
  const topicFilter = ct => ct.topic.name.toLowerCase().indexOf(searchTerm?.toLowerCase()) > -1
  const topicSort = (a, b) => {
    if (a.isSubscribed && !b.isSubscribed) return -1
    if (!a.isSubscribed && b.isSubscribed) return 1
    return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1
  }
  const allTopics = getGroupTopics(state, queryResultParams)
  const groupHasTopics = allTopics.length > 0
  const filteredTopics = allTopics
    .filter(topicFilter)
    .map(presentGroupTopic)
    .sort(topicSort)

  return {
    pending,
    group,
    groupHasTopics,
    filteredTopics,
    searchTerm
  }
}

export const mapDispatchToProps = {
  fetchGroupTopics,
  setTopicSubscribe,
  setTerm
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { group } = stateProps
  const { navigation } = ownProps
  const groupId = group?.id
  const fetchGroupTopics = () =>
    dispatchProps.fetchGroupTopics(groupId, { first: null })
  const setTopicSubscribe = (topicId, isSubscribing) =>
    dispatchProps.setTopicSubscribe(topicId, groupId, isSubscribing)
  const goToTopic = topicName => navigation.navigate('Topic Feed', { topicName })

  return {
    ...ownProps,
    ...stateProps,
    ...dispatchProps,
    fetchGroupTopics,
    setTopicSubscribe,
    goToTopic
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
