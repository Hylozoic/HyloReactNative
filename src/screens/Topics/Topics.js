import React, { useRef, useCallback } from 'react'
import { Text, ScrollView, Image, View, TouchableOpacity } from 'react-native'
import { useFocusEffect, useIsFocused, useScrollToTop } from '@react-navigation/native'
import KeyboardFriendlyView from 'components/KeyboardFriendlyView'
import Loading from 'components/Loading'
import StarIcon from 'components/StarIcon'
import Badge from 'components/Badge'
import SearchBar from 'components/SearchBar'
import styles from './Topics.styles'
import EntypoIcon from 'react-native-vector-icons/Entypo'
import LinearGradient from 'react-native-linear-gradient'
import { isEmpty, get } from 'lodash/fp'
import { bannerlinearGradientColors } from 'style/colors'

export default function Topics ({
  groupHasTopics,
  group,
  filteredTopics,
  pending,
  setTopicSubscribe,
  goToTopic,
  searchTerm,
  setTerm,
  fetchGroupTopics
}) {
  const ref = useRef(null)

  useScrollToTop(ref)
  useFocusEffect(
    useCallback(() => {
      group?.id && fetchGroupTopics()
    }, [group.id])
  )

  const bannerUrl = get('bannerUrl', group)
  const name = get('name', group)
  const image = { uri: bannerUrl }

  return (
    <KeyboardFriendlyView style={styles.container}>
      <ScrollView ref={ref}>
        <View style={styles.bannerContainer}>
          {image && <Image source={image} style={styles.image} />}
          <LinearGradient style={styles.gradient} colors={bannerlinearGradientColors} />
          <View style={styles.titleRow}>
            <Text style={styles.name}>{name}</Text>
          </View>
        </View>
        <SearchBar
          style={styles.searchBar}
          value={searchTerm}
          onChangeText={setTerm}
          placeholder='Search Topics'
          disable={!groupHasTopics}
        />
        {pending && isEmpty(filteredTopics)
          ? <Loading />
          : <TopicList
              groupHasTopics={groupHasTopics}
              searchTerm={searchTerm}
              filteredTopics={filteredTopics}
              setTopicSubscribe={setTopicSubscribe}
              goToTopic={goToTopic}
            />}
      </ScrollView>
    </KeyboardFriendlyView>
  )
}

export function TopicList ({
  groupHasTopics,
  filteredTopics,
  setTopicSubscribe,
  searchTerm,
  goToTopic
}) {
  if (!groupHasTopics) return (
    <View style={styles.topicList}>
      <Text style={styles.emptyList}>No topics were found for this group</Text>
    </View>
  )
  return (
    <View style={styles.topicList}>
      {isEmpty(filteredTopics) && searchTerm && (
        <Text style={styles.emptyList}>No topics matched "{searchTerm}"</Text>
      )}
      {!isEmpty(filteredTopics) && filteredTopics.map((topic, i) => (
        <TopicRow
          topic={topic} key={i} first={i === 0}
          setTopicSubscribe={setTopicSubscribe}
          goToTopic={goToTopic}
        />
      ))}
    </View>
  )
}

export function TopicRow ({ topic, first, setTopicSubscribe, goToTopic }) {
  const { name, newPostCount, isSubscribed } = topic
  return (
    <TouchableOpacity
      style={[styles.topicRow, first && styles.firstRow]}
      onPress={() => goToTopic(topic.name)}
    >
      <SubscribeStar isSubscribed={isSubscribed} onPress={() => setTopicSubscribe(topic.id, !isSubscribed)} />
      <Text style={styles.topicName}>#{name}</Text>
      <View style={styles.rightItems}>
        <Badge style={styles.badge} count={newPostCount} />
        <EntypoIcon style={styles.chevron} name='chevron-right' />
      </View>
    </TouchableOpacity>
  )
}

export function SubscribeStar ({ isSubscribed, onPress }) {
  const theme = isSubscribed
    ? styles.subscribedStar
    : styles.unSubscribedStar
  return (
    <TouchableOpacity
      style={styles.star}
      onPress={() => onPress(!isSubscribed)}
      hitSlop={{ top: 10, bottom: 10, left: 15, right: 15 }}
    >
      <StarIcon theme={theme} />
    </TouchableOpacity>
  )
}
