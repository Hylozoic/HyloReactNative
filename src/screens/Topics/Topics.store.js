import { makeGetQueryResults, makeQueryResultsModelSelector } from 'store/reducers/queryResults'
import { FETCH_GROUP_TOPICS } from 'store/actions/fetchGroupTopics'
export { setTopicSubscribe } from 'screens/Feed/Feed.store'

export const MODULE_NAME = 'Topics'
export const SET_TERM = `${MODULE_NAME}/SET_TERM`

const getGroupTopicResults = makeGetQueryResults(FETCH_GROUP_TOPICS)

export const getGroupTopics = makeQueryResultsModelSelector(
  getGroupTopicResults,
  'GroupTopic')

export function presentGroupTopic (groupTopic) {
  return {
    ...groupTopic.topic.ref,
    isSubscribed: groupTopic.isSubscribed,
    newPostCount: groupTopic.newPostCount
  }
}

export const defaultState = {
  term: ''
}

export default function reducer (state = defaultState, action) {
  const { error, type, payload } = action
  if (error) return state

  switch (type) {
    case SET_TERM:
      return {
        ...state,
        term: payload
      }
    default:
      return state
  }
}

export function setTerm (term) {
  return {
    type: SET_TERM,
    payload: term
  }
}

export function getTerm (state) {
  return state[MODULE_NAME].term
}
