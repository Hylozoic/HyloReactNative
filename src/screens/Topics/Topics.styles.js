import { StyleSheet } from 'react-native'
import { white, caribbeanGreen, capeCod40, capeCod20, rhino50, ghost } from 'style/colors'
import { isIOS } from 'util/platform'

const hasTextShadow = {
  textShadowColor: 'rgba(0, 0, 0, 0.25)',
  textShadowOffset: { width: 0, height: 2 },
  textShadowRadius: 7
}

export default {
  container: {
    backgroundColor: white,
    flex: 1
  },
  bannerContainer: {
    zIndex: 10,
    height: 142,
    marginBottom: 10
  },
  image: {
    height: 142,
    width: '100%',
    position: 'absolute'
  },
  gradient: {
    height: 142,
    width: '100%',
    position: 'absolute'
  },
  titleRow: {
    position: 'absolute',
    left: 0,
    bottom: 56,
    marginHorizontal: 16,
    flexDirection: 'row',
    alignItems: 'flex-end'
  },
  name: {
    fontSize: 24,
    fontFamily: 'Circular-Black',
    color: white,
    backgroundColor: 'transparent',
    ...hasTextShadow
  },
  // 
  searchBar: {
    marginHorizontal: 8,
    marginBottom: 12
  },
  topicList: {
    borderTopWidth: StyleSheet.hairlineWidth,
    borderTopColor: capeCod40,
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: capeCod40,
    paddingLeft: 15
  },
  emptyList: {
    paddingVertical: 10
  },
  topicRow: {
    borderTopWidth: StyleSheet.hairlineWidth,
    borderTopColor: capeCod40,
    height: 44,
    flexDirection: 'row',
    alignItems: 'center'
  },
  firstRow: {
    borderTopWidth: 0
  },
  topicName: {
    fontSize: 16
  },
  rightItems: {
    flexDirection: 'row',
    marginLeft: 'auto'
  },
  badge: {
    marginRight: 10,
    position: 'relative',
    top: 3
  },
  chevron: {
    marginRight: 12,
    fontSize: 24,
    color: capeCod20
  },
  star: {
    marginRight: 15
  },
  // themes
  unSubscribedStar: {
    icon: {
      fontSize: 30,
      color: capeCod20,
      backgroundColor: 'white'
    }
  },
  subscribedStar: {
    icon: {
      fontSize: 30,
      color: caribbeanGreen,
      backgroundColor: 'transparent'
    }
  }
}
