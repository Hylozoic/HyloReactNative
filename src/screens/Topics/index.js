import component from './Topics'
import connector from './Topics.connector'

export default connector(component)
