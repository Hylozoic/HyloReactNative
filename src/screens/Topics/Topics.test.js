import 'react-native'
import React from 'react'
import { render } from '@testing-library/react-native'
import ReactShallowRenderer from 'react-test-renderer/shallow'
import Topics, { TopicList, TopicRow, SubscribeStar } from './Topics'
import './TopicSupportComingSoon'
import '@react-navigation/native'

jest.mock('./TopicSupportComingSoon', () => 'TopicSupportComingSoon')
jest.mock('@react-navigation/native')

describe('Topics', () => {
  describe('"Support Coming Soon"', () => {
    it('displayed when no community id', () => {
      const props = {
        fetchCommunityTopics: jest.fn()
      }
      const { toJSON } = render(
        <Topics {...props} />
      )
      expect(props.fetchCommunityTopics).not.toBeCalled()
      expect(toJSON()).toMatchSnapshot()
    })

    it("not displayed when a community id is present, fetches", () => {
      const props = {
        community: { id: 123 },
        fetchCommunityTopics: jest.fn()
      }
      const { toJSON } = render(
        <Topics {...props} />
      )
      expect(props.fetchCommunityTopics).toBeCalled()
      expect(toJSON()).toMatchSnapshot()
    })

    it('fetches when rerendered with a communityId', () => {
      const propsBefore = {
        fetchCommunityTopics: jest.fn()
      }
      const propsAfter = {
        community: { id: 123 },
        fetchCommunityTopics: jest.fn()
      }
      const { rerender } = render(
        <Topics {...propsBefore} />
      )
      rerender(
        <Topics {...propsAfter} />
      )
      expect(propsBefore.fetchCommunityTopics).not.toBeCalled()
      expect(propsAfter.fetchCommunityTopics).toBeCalled()
    })
  })

  it('renders correctly a community and topics', () => {
    const props = {
      fetchCommunityTopics: () => {},
      community: {
        id: 123,
        name: 'Foomunity',
        bannerUrl: 'community-banner.png'
      },
      topics: [1, 2, 3],
      setTopicSubscribe: () => {},
      goToTopic: () => {}
    }
    const { toJSON } = render(
      <Topics {...props} />
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('renders correctly with pending=true', () => {
    const props = {
      fetchCommunityTopics: () => {},
      community: {
        id: 123,
        name: 'Foomunity',
        bannerUrl: 'community-banner.png'
      },
      topics: [1, 2, 3],
      pending: true,
      setTopicSubscribe: () => {},
      goToTopic: () => {}
    }
    const { toJSON } = render(
      <Topics {...props} />
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('refetches when rerendered with a different communityId', () => {
    const propsBefore = {
      community: { id: 333 },
      fetchCommunityTopics: jest.fn()
    }
    const propsAfter = {
      community: { id: 222 },
      fetchCommunityTopics: jest.fn()
    }
    const { rerender } = render(
      <Topics {...propsBefore} />
    )
    rerender(
      <Topics {...propsAfter} />
    )
    expect(propsBefore.fetchCommunityTopics).toBeCalled()
    expect(propsAfter.fetchCommunityTopics).toBeCalled()
  })
})

describe('TopicList', () => {
  it('matches last snapshot', () => {
    const props = {
      topics: [1, 2, 3],
      setTopicSubscribe: () => {},
      goToTopic: () => {}
    }
    const renderer = new ReactShallowRenderer()
    renderer.render(<TopicList {...props} />)
    const actual = renderer.getRenderOutput()
    expect(actual).toMatchSnapshot()
  })

  it('matches last snapshot with empty list', () => {
    const props = {
      topics: [],
      setTopicSubscribe: () => {},
      goToTopic: () => {}
    }
    const renderer = new ReactShallowRenderer()
    renderer.render(<TopicList {...props} />)
    const actual = renderer.getRenderOutput()
    expect(actual).toMatchSnapshot()
  })
})

describe('TopicRow', () => {
  it('matches last snapshot', () => {
    const props = {
      topic: {
        name: 'theTopic',
        newPostCount: 4,
        isSubscribed: true
      },
      setTopicSubscribe: () => {},
      goToTopic: () => {}
    }
    const renderer = new ReactShallowRenderer()
    renderer.render(<TopicRow {...props} />)
    const actual = renderer.getRenderOutput()
    expect(actual).toMatchSnapshot()
  })
})

describe('SubscribeStar', () => {
  it('matches last snapshot', () => {
    const props = {
      isSubscribed: true,
      onPress: () => {}
    }
    const renderer = new ReactShallowRenderer()
    renderer.render(<SubscribeStar {...props} />)
    const actual = renderer.getRenderOutput()
    expect(actual).toMatchSnapshot()
  })
})
