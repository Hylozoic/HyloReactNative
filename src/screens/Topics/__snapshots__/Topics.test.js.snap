// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscribeStar matches last snapshot 1`] = `
<ForwardRef
  hitSlop={
    Object {
      "bottom": 10,
      "left": 15,
      "right": 15,
      "top": 10,
    }
  }
  onPress={[Function]}
>
  <StarIcon
    theme={
      Object {
        "icon": Object {
          "backgroundColor": "transparent",
          "color": "#0DC39F",
          "fontSize": 30,
        },
      }
    }
  />
</ForwardRef>
`;

exports[`TopicList matches last snapshot 1`] = `
<View>
  <Text>
    No topics were found for this community
  </Text>
</View>
`;

exports[`TopicList matches last snapshot with empty list 1`] = `
<View>
  <Text>
    No topics were found for this community
  </Text>
</View>
`;

exports[`TopicRow matches last snapshot 1`] = `
<ForwardRef
  onPress={[Function]}
>
  <SubscribeStar
    isSubscribed={true}
    onPress={[Function]}
  />
  <Text>
    #
    theTopic
  </Text>
  <View>
    <Badge
      count={4}
    />
    <Icon
      allowFontScaling={false}
      name="chevron-right"
      size={12}
    />
  </View>
</ForwardRef>
`;

exports[`Topics "Support Coming Soon" displayed when no community id 1`] = `<TopicSupportComingSoon />`;

exports[`Topics "Support Coming Soon" not displayed when a community id is present, fetches 1`] = `
<View
  onLayout={[Function]}
>
  <RCTScrollView>
    <View>
      <View>
        <Image
          source={
            Object {
              "uri": undefined,
            }
          }
        />
        <div
          colors={
            Array [
              "rgba(0, 0, 0, 0)",
              "rgba(0, 0, 0, 0.1)",
              "rgba(0, 0, 0, 0.3)",
              "rgba(0, 0, 0, 0.6)",
            ]
          }
        >
          MockLinearGradient
        </div>
        <View>
          <Text />
        </View>
      </View>
      <View>
        <Icon
          name="Search"
        />
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          autoFocus={false}
          editable={true}
          placeholder="Search Topics"
          rejectResponderTermination={true}
          underlineColorAndroid="transparent"
        />
      </View>
      <View>
        <Text>
          No topics were found for this community
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Topics renders correctly a community and topics 1`] = `
<View
  onLayout={[Function]}
>
  <RCTScrollView>
    <View>
      <View>
        <Image
          source={
            Object {
              "uri": "community-banner.png",
            }
          }
        />
        <div
          colors={
            Array [
              "rgba(0, 0, 0, 0)",
              "rgba(0, 0, 0, 0.1)",
              "rgba(0, 0, 0, 0.3)",
              "rgba(0, 0, 0, 0.6)",
            ]
          }
        >
          MockLinearGradient
        </div>
        <View>
          <Text>
            Foomunity
          </Text>
        </View>
      </View>
      <View>
        <Icon
          name="Search"
        />
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          autoFocus={false}
          editable={true}
          placeholder="Search Topics"
          rejectResponderTermination={true}
          underlineColorAndroid="transparent"
        />
      </View>
      <View>
        <Text>
          No topics were found for this community
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Topics renders correctly with pending=true 1`] = `
<View
  onLayout={[Function]}
>
  <RCTScrollView>
    <View>
      <View>
        <Image
          source={
            Object {
              "uri": "community-banner.png",
            }
          }
        />
        <div
          colors={
            Array [
              "rgba(0, 0, 0, 0)",
              "rgba(0, 0, 0, 0.1)",
              "rgba(0, 0, 0, 0.3)",
              "rgba(0, 0, 0, 0.6)",
            ]
          }
        >
          MockLinearGradient
        </div>
        <View>
          <Text>
            Foomunity
          </Text>
        </View>
      </View>
      <View>
        <Icon
          name="Search"
        />
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          autoFocus={false}
          editable={true}
          placeholder="Search Topics"
          rejectResponderTermination={true}
          underlineColorAndroid="transparent"
        />
      </View>
      <View>
        <ActivityIndicator
          animating={true}
          color="#818A88"
          hidesWhenStopped={true}
          size="small"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
